File: implementation/csharp/GameEventScript.SyntaxHighlighter/src/GameEventScriptSyntaxHighlighter.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;

namespace GameEventScript.SyntaxHighlighter;

/// <summary>Reusable GES/GESA TextMate highlighting without a Compiler, Runtime, UI or terminal dependency.</summary>
public sealed class GameEventScriptSyntaxHighlighter
{
    private static readonly GrammarRule[] Rules = EmbeddedGrammars.CreateRules();
    /// <summary>The selected bundled grammar.</summary>
    public GameEventScriptSyntaxLanguage Language { get; }
    /// <summary>Maximum UTF-16 input length for one call; larger inputs return an incomplete result.</summary>
    public int MaxInputLength { get; }
    private readonly int[] _roots;
    private readonly string _rootScope;

    /// <summary>Creates a highlighter. Matching has a cooperative 1000ms per-call budget, plus a 100ms per-regex bound.</summary>
    /// <exception cref="ArgumentOutOfRangeException">The language is unknown or maxInputLength is not positive.</exception>
    public GameEventScriptSyntaxHighlighter(GameEventScriptSyntaxLanguage language = GameEventScriptSyntaxLanguage.Ges, int maxInputLength = 262144)
    {
        if (language is not (GameEventScriptSyntaxLanguage.Ges or GameEventScriptSyntaxLanguage.Gesa)) throw new ArgumentOutOfRangeException(nameof(language));
        if (maxInputLength <= 0) throw new ArgumentOutOfRangeException(nameof(maxInputLength));
        Language = language;
        MaxInputLength = maxInputLength;
        _roots = language == GameEventScriptSyntaxLanguage.Ges ? EmbeddedGrammars.Source : EmbeddedGrammars.Assembly;
        _rootScope = language == GameEventScriptSyntaxLanguage.Ges ? "source.gameeventscript" : "source.gameeventscript.assembler";
    }

    /// <summary>Creates an empty line-boundary state. States are immutable and can be retained by editors.</summary>
    public GameEventScriptHighlightState CreateState() => new(Language, []);

    /// <summary>Highlights a complete document, preserving CR, LF, CRLF and all UTF-16 offsets. Invalid source remains highlightable.</summary>
    /// <exception cref="ArgumentNullException">text is null.</exception>
    public GameEventScriptHighlightResult Highlight(string text) => Run(text, null, false);

    /// <summary>Highlights one line with an optional trailing CR, LF or CRLF. Offsets are relative to this line.</summary>
    /// <remarks>Supply the previous line's successful NextState. Reprocess subsequent lines until text and state converge.</remarks>
    /// <exception cref="ArgumentException">The text contains an interior line ending, or the state belongs to another language.</exception>
    /// <exception cref="ArgumentNullException">text is null.</exception>
    public GameEventScriptHighlightResult HighlightLine(string text, GameEventScriptHighlightState? state = null) => Run(text, state, true);

    private GameEventScriptHighlightResult Run(string text, GameEventScriptHighlightState? state, bool singleLine)
    {
        if (text is null) throw new ArgumentNullException(nameof(text));
        if (state is not null && state.Language != Language) throw new ArgumentException("State belongs to another language.", nameof(state));
        if (singleLine && text.Substring(0, ContentLength(text)).IndexOfAny(['\r', '\n']) >= 0) throw new ArgumentException("Expected one line.", nameof(text));
        if (text.Length > MaxInputLength) return new([], null);
        var clock = Stopwatch.StartNew();
        var frames = new List<int>(state?.Frames ?? []);
        var spans = new List<GameEventScriptHighlightSpan>();
        try
        {
            var start = 0;
            do
            {
                var end = start;
                while (end < text.Length && text[end] is not ('\r' or '\n')) end++;
                if (end < text.Length && text[end++] == '\r' && end < text.Length && text[end] == '\n') end++;
                var line = text.Substring(start, end - start);
                if (!Line(line, start, frames, spans, clock)) return new([], null);
                start = end;
            } while (start < text.Length);
            return new(spans, new(Language, frames));
        }
        catch (RegexMatchTimeoutException) { return new([], null); }
    }

    private bool Line(string text, int offset, List<int> frames, List<GameEventScriptHighlightSpan> spans, Stopwatch clock)
    {
        var length = ContentLength(text);
        var content = text.Substring(0, length);
        var cache = new Dictionary<int, Match>();
        Match Find(int key, Regex regex, int position)
        {
            if (!cache.TryGetValue(key, out var value) || value.Success && value.Index < position)
                cache[key] = value = regex.Match(content, position);
            return value;
        }
        var position = 0;
        while (position <= length)
        {
            if (clock.ElapsedMilliseconds > 1000 || frames.Count > 128) return false;
            var scopes = Context(frames);
            var parent = frames.Count == 0 ? -1 : frames[frames.Count - 1];
            Match? match = parent < 0 ? null : Find(parent * 2 + 1, Rules[parent].End!, position);
            if (match is not null && !match.Success) match = null;
            var ruleId = parent;
            var ending = match is not null;
            foreach (var id in parent < 0 ? _roots : Rules[parent].Children)
            {
                var candidate = Find(id * 2, Rules[id].Pattern, position);
                if (candidate.Success && (match is null || candidate.Index < match.Index || candidate.Index == match.Index && ending && Rules[parent].EndLast))
                {
                    match = candidate;
                    ruleId = id;
                    ending = false;
                }
            }
            if (match is null)
            {
                Add(spans, offset + position, length - position, scopes);
                break;
            }
            Add(spans, offset + position, match.Index - position, scopes);
            var rule = Rules[ruleId];
            if (ending)
            {
                if (rule.ContentName.Length != 0) scopes.RemoveAt(scopes.Count - 1);
                Emit(spans, offset, match, scopes, rule.EndCaptures);
                frames.RemoveAt(frames.Count - 1);
            }
            else
            {
                if (match.Length == 0) return false;
                if (rule.Name.Length != 0) scopes.Add(rule.Name);
                Emit(spans, offset, match, scopes, rule.End is null ? rule.Captures : rule.BeginCaptures);
                if (rule.End is not null) frames.Add(ruleId);
            }
            position = match.Index + match.Length;
        }
        Add(spans, offset + length, text.Length - length, Context(frames));
        return clock.ElapsedMilliseconds <= 1000;
    }

    private List<string> Context(List<int> frames)
    {
        var scopes = new List<string> { _rootScope };
        foreach (var id in frames)
        {
            if (Rules[id].Name.Length != 0) scopes.Add(Rules[id].Name);
            if (Rules[id].ContentName.Length != 0) scopes.Add(Rules[id].ContentName);
        }
        return scopes;
    }

    private static void Emit(List<GameEventScriptHighlightSpan> spans, int offset, Match match, List<string> scopes, CaptureRule[] captures)
    {
        var groups = captures.Where(c => c.Group < match.Groups.Count && match.Groups[c.Group].Success && match.Groups[c.Group].Length > 0)
            .Select(c => (Range: match.Groups[c.Group], c.Scope)).OrderByDescending(c => c.Range.Length).ToArray();
        var boundaries = new SortedSet<int> { match.Index, match.Index + match.Length };
        foreach (var group in groups)
        {
            boundaries.Add(group.Range.Index);
            boundaries.Add(group.Range.Index + group.Range.Length);
        }
        var points = boundaries.ToArray();
        for (var i = 0; i + 1 < points.Length; i++)
        {
            var names = new List<string>(scopes);
            foreach (var group in groups)
                if (group.Range.Index <= points[i] && points[i] < group.Range.Index + group.Range.Length) names.Add(group.Scope);
            Add(spans, offset + points[i], points[i + 1] - points[i], names);
        }
    }

    private static void Add(List<GameEventScriptHighlightSpan> spans, int start, int length, List<string> scopes)
    {
        if (length <= 0) return;
        if (spans.Count > 0)
        {
            var previous = spans[spans.Count - 1];
            if (previous.Start + previous.Length == start && previous.Scopes.SequenceEqual(scopes))
            {
                spans[spans.Count - 1] = new(previous.Start, previous.Length + length, scopes.ToArray());
                return;
            }
        }
        spans.Add(new(start, length, scopes.ToArray()));
    }

    private static int ContentLength(string text)
    {
        var length = text.Length;
        if (length > 0 && text[length - 1] == '\n') length--;
        if (length > 0 && text[length - 1] == '\r') length--;
        return length;
    }
}

internal sealed class CaptureRule(int group, string scope)
{
    internal int Group { get; } = group;
    internal string Scope { get; } = scope;
}

internal sealed class GrammarRule
{
    internal Regex Pattern { get; }
    internal Regex? End { get; }
    internal string Name { get; }
    internal string ContentName { get; }
    internal CaptureRule[] Captures { get; }
    internal CaptureRule[] BeginCaptures { get; }
    internal CaptureRule[] EndCaptures { get; }
    internal int[] Children { get; }
    internal bool EndLast { get; }

    internal GrammarRule(string pattern, string end, string name, string contentName, CaptureRule[] captures, CaptureRule[] beginCaptures, CaptureRule[] endCaptures, int[] children, bool endLast)
    {
        const RegexOptions options = RegexOptions.CultureInvariant | RegexOptions.Multiline;
        Pattern = new(pattern, options, TimeSpan.FromMilliseconds(100));
        End = end.Length == 0 ? null : new(end, options, TimeSpan.FromMilliseconds(100));
        Name = name;
        ContentName = contentName;
        Captures = captures;
        BeginCaptures = beginCaptures;
        EndCaptures = endCaptures;
        Children = children;
        EndLast = endLast;
    }
}

File: implementation/swift/GameEventScriptSyntaxHighlighter/Sources/GameEventScriptSyntaxHighlighter/GameEventScriptSyntaxHighlighter.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// Reusable TextMate highlighting without a Compiler, Runtime, UI or terminal dependency. Matching uses Foundation internally.
public final class GameEventScriptSyntaxHighlighter: Sendable {
    /// The selected bundled grammar.
    public let language: GameEventScriptSyntaxLanguage
    /// Maximum UTF-16 input length per call; larger inputs return an incomplete result.
    public let maxInputLength: Int
    private let roots: [Int]
    private let rootScope: String

    /// Creates a highlighter with a cooperative 1000ms per-call matching budget.
    /// - Throws: `invalidInputLimit` if the bound is not positive. Foundation progress checks are not a hard real-time deadline.
    public init(language: GameEventScriptSyntaxLanguage = .ges, maxInputLength: Int = 262_144) throws {
        guard maxInputLength > 0 else { throw GameEventScriptHighlightingError.invalidInputLimit }
        self.language = language
        self.maxInputLength = maxInputLength
        roots = language == .ges ? Grammar.shared.source : Grammar.shared.assembly
        rootScope = language == .ges ? "source.gameeventscript" : "source.gameeventscript.assembler"
    }

    /// Creates an empty, immutable line-boundary state that editors may retain.
    public func createState() -> GameEventScriptHighlightState { .init(language: language, frames: []) }

    /// Highlights a complete document, preserving CR, LF, CRLF and all UTF-16 offsets. Invalid code remains highlightable.
    public func highlight(_ text: String) -> GameEventScriptHighlightResult { run(text, state: nil) }

    /// Highlights one line with an optional trailing CR, LF or CRLF; offsets are relative to this line.
    /// Supply the previous line's successful nextState and reprocess subsequent lines until text and state converge.
    /// - Throws: `expectedSingleLine` for interior line endings, or `incompatibleState` for a different grammar.
    public func highlightLine(_ text: String, state: GameEventScriptHighlightState? = nil) throws -> GameEventScriptHighlightResult {
        if let state, state.language != language { throw GameEventScriptHighlightingError.incompatibleState }
        let source = text as NSString
        let content = source.substring(to: Self.contentLength(source))
        if content.contains("\r") || content.contains("\n") { throw GameEventScriptHighlightingError.expectedSingleLine }
        return run(text, state: state)
    }

    private func run(_ text: String, state: GameEventScriptHighlightState?) -> GameEventScriptHighlightResult {
        guard text.utf16.count <= maxInputLength else { return .init(spans: [], nextState: nil) }
        let deadline = ProcessInfo.processInfo.systemUptime + 1.0
        let source = text as NSString
        var frames = state?.frames ?? []
        var spans: [GameEventScriptHighlightSpan] = []
        var start = 0
        repeat {
            var end = start
            while end < source.length && source.character(at: end) != 13 && source.character(at: end) != 10 { end += 1 }
            if end < source.length {
                let cr = source.character(at: end) == 13
                end += 1
                if cr && end < source.length && source.character(at: end) == 10 { end += 1 }
            }
            let text = source.substring(with: NSRange(location: start, length: end - start))
            guard line(text, offset: start, frames: &frames, spans: &spans, deadline: deadline) else { return .init(spans: [], nextState: nil) }
            start = end
        } while start < source.length
        return .init(spans: spans, nextState: .init(language: language, frames: frames))
    }

    private func line(_ text: String, offset: Int, frames: inout [Int], spans: inout [GameEventScriptHighlightSpan], deadline: Double) -> Bool {
        let source = text as NSString
        let length = Self.contentLength(source)
        let content = source.substring(to: length)
        var cache: [Int: CachedMatch] = [:]
        var expired = false

        func find(_ key: Int, _ regex: NSRegularExpression, _ position: Int) -> NSTextCheckingResult? {
            if let previous = cache[key], previous.match == nil || previous.match!.range.location >= position { return previous.match }
            var result: NSTextCheckingResult?
            regex.enumerateMatches(in: content, options: [.reportProgress, .withTransparentBounds, .withoutAnchoringBounds], range: NSRange(location: position, length: length - position)) { match, _, stop in
                if ProcessInfo.processInfo.systemUptime > deadline {
                    expired = true
                    stop.pointee = true
                } else if let match {
                    result = match
                    stop.pointee = true
                }
            }
            cache[key] = CachedMatch(match: result)
            return result
        }

        let rules = Grammar.shared.rules
        var position = 0
        while position <= length {
            if expired || ProcessInfo.processInfo.systemUptime > deadline || frames.count > 128 { return false }
            var scopes = context(frames)
            let parent = frames.last
            var match = parent.flatMap { find($0 * 2 + 1, rules[$0].end!, position) }
            var ruleID = parent ?? 0
            var ending = match != nil
            for id in parent.map({ rules[$0].children }) ?? roots {
                if let candidate = find(id * 2, rules[id].pattern, position),
                    match == nil || candidate.range.location < match!.range.location || candidate.range.location == match!.range.location && ending && rules[parent!].endLast
                {
                    match = candidate
                    ruleID = id
                    ending = false
                }
            }
            guard let match else {
                Self.add(&spans, offset + position, length - position, scopes)
                break
            }
            Self.add(&spans, offset + position, match.range.location - position, scopes)
            let rule = rules[ruleID]
            if ending {
                if !rule.contentName.isEmpty { scopes.removeLast() }
                Self.emit(&spans, offset, match, scopes, rule.endCaptures)
                frames.removeLast()
            } else {
                if match.range.length == 0 { return false }
                if !rule.name.isEmpty { scopes.append(rule.name) }
                Self.emit(&spans, offset, match, scopes, rule.end == nil ? rule.captures : rule.beginCaptures)
                if rule.end != nil { frames.append(ruleID) }
            }
            position = NSMaxRange(match.range)
        }
        Self.add(&spans, offset + length, source.length - length, context(frames))
        return !expired && ProcessInfo.processInfo.systemUptime <= deadline
    }

    private func context(_ frames: [Int]) -> [String] {
        var scopes = [rootScope]
        for id in frames {
            let rule = Grammar.shared.rules[id]
            if !rule.name.isEmpty { scopes.append(rule.name) }
            if !rule.contentName.isEmpty { scopes.append(rule.contentName) }
        }
        return scopes
    }

    private static func emit(_ spans: inout [GameEventScriptHighlightSpan], _ offset: Int, _ match: NSTextCheckingResult, _ scopes: [String], _ captures: [CaptureRule]) {
        let groups = captures.compactMap { capture -> (NSRange, String, Int)? in
            guard capture.group < match.numberOfRanges else { return nil }
            let range = match.range(at: capture.group)
            return range.location == NSNotFound || range.length == 0 ? nil : (range, capture.scope, capture.group)
        }.sorted { $0.0.length != $1.0.length ? $0.0.length > $1.0.length : $0.2 < $1.2 }
        var boundaries: Set<Int> = [match.range.location, NSMaxRange(match.range)]
        for (range, _, _) in groups { boundaries.formUnion([range.location, NSMaxRange(range)]) }
        let points = boundaries.sorted()
        for (start, end) in zip(points, points.dropFirst()) {
            let names = scopes + groups.filter { NSLocationInRange(start, $0.0) }.map { $0.1 }
            add(&spans, offset + start, end - start, names)
        }
    }

    private static func add(_ spans: inout [GameEventScriptHighlightSpan], _ start: Int, _ length: Int, _ scopes: [String]) {
        guard length > 0 else { return }
        if let previous = spans.last, previous.start + previous.length == start, previous.scopes == scopes {
            spans[spans.count - 1] = .init(previous.start, previous.length + length, scopes)
        } else {
            spans.append(.init(start, length, scopes))
        }
    }

    private static func contentLength(_ text: NSString) -> Int {
        var length = text.length
        if length > 0 && text.character(at: length - 1) == 10 { length -= 1 }
        if length > 0 && text.character(at: length - 1) == 13 { length -= 1 }
        return length
    }
}

private struct CachedMatch { let match: NSTextCheckingResult? }

struct CaptureRule {
    let group: Int
    let scope: String
}

final class GrammarRule {
    let pattern: NSRegularExpression
    let end: NSRegularExpression?
    let name: String
    let contentName: String
    let captures: [CaptureRule]
    let beginCaptures: [CaptureRule]
    let endCaptures: [CaptureRule]
    let children: [Int]
    let endLast: Bool

    init(_ data: [String: Any]) {
        // Physical CR/LF lines are split before matching. Other Unicode separators
        // must remain ordinary content, matching .NET Regex's LF-only line semantics.
        let options: NSRegularExpression.Options = [.anchorsMatchLines, .useUnixLineSeparators]
        pattern = try! NSRegularExpression(pattern: data["pattern"] as! String, options: options)
        let ending = data["end"] as! String
        end = ending.isEmpty ? nil : try! NSRegularExpression(pattern: ending, options: options)
        name = data["name"] as! String
        contentName = data["contentName"] as! String

        func groups(_ key: String) -> [CaptureRule] { (data[key] as! [[Any]]).map { CaptureRule(group: $0[0] as! Int, scope: $0[1] as! String) } }

        captures = groups("captures")
        beginCaptures = groups("beginCaptures")
        endCaptures = groups("endCaptures")
        children = data["children"] as! [Int]
        endLast = data["endLast"] as! Bool
    }
}

// Immutable compiled rules. NSRegularExpression is immutable; matching state remains local to each call.
final class Grammar: @unchecked Sendable {
    static let shared = Grammar()
    let source: [Int]
    let assembly: [Int]
    let rules: [GrammarRule]

    private init() {
        let data = try! JSONSerialization.jsonObject(with: Data(EmbeddedGrammars.json.utf8)) as! [String: Any]
        source = data["source"] as! [Int]
        assembly = data["assembly"] as! [Int]
        rules = (data["rules"] as! [[String: Any]]).map(GrammarRule.init)
    }
}
